[HLSL][RootSignature] Implement Lexing of DescriptorTables

inbelic opened the pull request:
For the sake of scope, we will let the lexing of floating literals be deferred until needed for Static Samplers. Other than that this pr should allow us to simply define new enumerations/keywords in `RootSignatureTokenKinds.def` for when they are used in the parser. We could have defined all of these keywords here, but for the sake of correctness in review we will let them be split up.

- Define `RootSignatureLexer` and provide a public `LexToken` method for external use
- Define the file `RootSignatureTokenKinds` to define required tokens and allow for future custom keywords/enums
- Implement the internal methods required to parse the different types of tokens (integers, flag enums, puncuators...)
- Add test harness for unit testing and the respective unit tests for lexing the tokens

Resolves #126563

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,151 @@
+    return false;
+  }
+
+  // Keywords and Enums:
```
For the sake of readability all keywords and enums are defined into a single switch construct. However, it would be possible to match the first character to reduce the amount of string comparisons if performance is a concern.

For example:
```
if (toupper(C) == 'S') {
  Create a StringSwitch with only enums that start with 'S'
}
```

llvm-beanz replied:
I definitely prefer starting with the simple approach and only going complicated if we need to. DXC had a bunch of hand-rolled string matching functions that unrolled the strings... and a few of them had some bad bugs. So this is the perfect way to start.

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,152 @@
+  AdvanceBuffer(Buffer.take_while(isspace).size());
+
+  while (!Buffer.empty()) {
+    RootSignatureToken Result;
```
Switch to just using the `SourceLocation` constructor and remove the default constructor.

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
I have left the error reporting as a todo just to help with pr size, as it will require adding a good chunk of additional testing

llvm-beanz replied:
As I said in your other PR, I believe not including error reporting and the appropriate testing makes it significantly more difficult to review the subsequent commit since not all the appropriate code will even be in the PR.

inbelic commented on clang/include/clang/Parse/HLSLRootSignatureTokenKinds.def:
```diff
@@ -0,0 +1,121 @@
+KEYWORD(offset)
+
+// Descriptor Range Offset Enum:
+DESCRIPTOR_RANGE_OFFSET_ENUM(DescriptorRangeOffsetAppend, "DESCRIPTOR_RANGE_OFFSET_APPEND")
```
I contemplated to have the different enums defined in a seperate `.def` file with the form of:
`ENUM(NAME, VAL, LIT)` so that we will just define them in a single spot instead of here and in `llvm/include/llvm/Frontend/HLSL/HLSLRootSignature.h` (from parser pr).

But I felt that having this `.def` include another `.def` file become more confusing than helpful, and having them all in a single file would mean having token logic where it isn't applicable. Would appreciate suggestions here.

V-FEXrt commented on clang/unittests/Parse/ParseHLSLRootSignatureTest.cpp:
```diff
@@ -0,0 +1,130 @@
+    HeaderSearch HeaderInfo(std::make_shared<HeaderSearchOptions>(), SourceMgr,
+                            Diags, LangOpts, Target.get());
+    Preprocessor *PP =
+        new Preprocessor(std::make_shared<PreprocessorOptions>(), Diags,
```
imo unless you have a strong reason for it, you should _really_ avoid new/delete and instead use C++ smart pointers for allocating/freeing heap memory. It takes care of the memory management for you and prevents accidental memory leaks.

In this case, `std::unique_ptr` would work perfectly https://en.cppreference.com/w/cpp/memory/unique_ptr

V-FEXrt replied:
I think you can just derefence the `unique_ptr`. Something like `*ptr->get()`

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,152 @@
+
+bool RootSignatureLexer::LexNumber(RootSignatureToken &Result) {
+  // NumericLiteralParser does not handle the sign so we will manually apply it
+  Result.Signed = Buffer.front() == '-';
```
I guess DXC actually does allow a `+` to be the first character, as well as, `.`. For the sake of compatibility, we should bring those forward.

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,152 @@
+    return true; // Error has already been reported so just return
+
+  // Retrieve the number value to store into the token
+  if (Literal.isIntegerLiteral()) {
+    Result.Kind = TokenKind::int_literal;
+
+    APSInt X = APSInt(32, Result.Signed);
+    if (Literal.GetIntegerValue(X))
+      return true; // TODO: Report overflow error
+
+    X = Result.Signed ? -X : X;
+    Result.IntLiteral = (uint32_t)X.getZExtValue();
+  } else {
+    return true; // TODO: report unsupported number literal specification
+  }
```
That follows the rest of the code conventions, thanks.

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,152 @@
+    AdvanceBuffer();                                                           \
+    return false;                                                              \
+  }
+#include "clang/Parse/HLSLRootSignatureTokenKinds.def"
```
Yes, it is defined this way so that the `PUNCUATOR` that is in the `.def` is the "default" macro. This is used when we do `#define TOK(X) X,` so then `PUNCUATOR(X) -> TOK(pu_X) -> X,`. It follows the pattern in some other `TokenKinds.def`.

inbelic commented on clang/include/clang/Parse/ParseHLSLRootSignature.h:
Note to self: change namespace from `llvm` to `clang`

V-FEXrt commented on clang/unittests/Parse/ParseHLSLRootSignatureTest.cpp:
```diff
@@ -0,0 +1,133 @@
+  SmallVector<RootSignatureToken> Tokens = {
+      RootSignatureToken() // invalid token for completeness
+  };
+  ASSERT_FALSE(Lexer.Lex(Tokens));
```
Would be nice to see a more robust set of failure tests

inbelic replied:
Yes, I decided to leave out the implementation of generating and testing diagnostics from this pr, just to help with the size. If you feel that should they be here to ensure correctness I could add it.

V-FEXrt replied:
No, I'm happy to punt it

llvm-beanz replied:
I'll reiterate, I think this is the wrong tradeoff, because it makes it significantly more difficult to review the follow-up change since we'll need to compare not just the incoming PR but also go back and look at this PR to ensure all the appropriate error states are captured and tested.

V-FEXrt commented on clang/unittests/Parse/ParseHLSLRootSignatureTest.cpp:
```diff
@@ -0,0 +1,133 @@
+  };
+  ASSERT_FALSE(Lexer.Lex(Tokens));
+
+  SmallVector<TokenKind> Expected = {
```
Its probably my unfamiliarity with the code but I can't see how this is actually building up the list of expected tokens. Importing the `.def` file here _seems_ strange

inbelic replied:
This test-case illustrates that we can lex all of the defined tokens in the `TokenKinds.def`, so to save typing all the token kinds out again I construct a vector of all the tokenkinds using a macro. It does so my creating a big initialization list for the SmallVector.

V-FEXrt reviewed (approved):
Personally happy with the code at a high level, though it would be best to have someone with domain knowledge also approve

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,153 @@
+// Lexer Definitions
+
+static bool IsNumberChar(char C) {
+  // TODO: extend for float support with or without hexadecimal/exponent
```
Yep, I agree. In this case, it will be added when I get to supporting `StaticSampler`s so it will be gone by the time the parent issue is resolved. I could update the todo to reference the parent issue though?

V-FEXrt replied:
Yep that would be great!

llvm-beanz commented on clang/include/clang/Parse/ParseHLSLRootSignature.h:
```diff
@@ -0,0 +1,96 @@
+  bool Signed = false;
+  union {
+    uint32_t IntLiteral = 0;
+    float FloatLiteral;
```
Should we have a union here or should this be an `llvm::APValue`?

damyanp commented on clang/unittests/Parse/ParseHLSLRootSignatureTest.cpp:
```diff
@@ -0,0 +1,134 @@
+      : FileMgr(FileMgrOpts), DiagID(new DiagnosticIDs()),
+        Diags(DiagID, new DiagnosticOptions, new IgnoringDiagConsumer()),
+        SourceMgr(Diags, FileMgr), TargetOpts(new TargetOptions) {
+    TargetOpts->Triple = "x86_64-apple-darwin11.1.0";
```
How was `"x86_64-apple-darwin11.1.0"` chosen as a value to use here?  Does it really mean "I don't care"?

inbelic replied:
It was taken over from https://github.com/llvm/llvm-project/blob/main/clang/unittests/Lex/LexerTest.cpp and it should denote "I don't care"

inbelic committed:
Add support for lexing registers

inbelic committed:
Add lexing for example keyword and enum

inbelic committed:
Add lexing for remaining DescriptorTable keywords and enums

inbelic committed:
[HLSL][RootSignature] Initial Lexer Definition with puncuators

- Defines the RootSignatureLexer class
- Defines the test harness required for testing

- Implements the punctuator tokens and tests functionality

inbelic committed:
Add lexing of integer literals

- Integrate the use of the `NumericLiteralParser` to lex integer
literals
- Add additional hlsl specific diagnostics messages

inbelic commented:
Updated to include diagnostics output and the relevant testing.

I have rebased to split the pr into smaller incremental changes. But I have taken care to address all previous comments.

inbelic commented:
HLSL Build failure is because there is an assertion that DIAG_SIZE_LEX is not big enough (insufficient). I did add some new diag errors under the Lex category so it seems the threshold is passed on some environments.

From [here](https://github.com/llvm/llvm-project/blame/617278e7b0c937fccbf7d67d14f053c3409bc33f/clang/include/clang/Basic/DiagnosticIDs.h#L40), it seems that we could, and there is trend to, bump this limit when needed. IIUC, I should create a separate pr to bump this.

inbelic committed:
review: update uses of llvm_unreachable

inbelic committed:
review: update target test triple with comment

inbelic committed:
review: dealing with signed positive integer

- the previous implementation would not have thrown an overflow error as
expected. Instead it would have been treated as a negative int
- DXC treats all positively signed ints as an unsigned integer, so for
compatibility, we will be doing the same.
- add better unit tests to demonstrate expected functionality

damyanp commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,171 @@
+// Lexer Definitions
+
+static bool IsNumberChar(char C) {
+  // TODO(#120472): extend for float support exponents
```
The issue linked here (#120472) is just about implementing the parser. Is there a reason not to get float tokens recognized by the lexer?

inbelic replied:
Updated the comment to link to https://github.com/llvm/llvm-project/issues/126565, which better describes this.

damyanp reviewed (commented):
LGTM - I'm surprised we don't have all of the tokens we're going to need for root signatures included in this PR.

llvm-beanz commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,171 @@
+
+  // All following tokens require at least one additional character
+  if (Buffer.size() <= 1) {
+    PP.getDiagnostics().Report(Result.TokLoc, diag::err_hlsl_invalid_token);
```
This error doesn't seem particularly useful to end users who make a mistake. Why not just return false here because it can't construct a token?

inbelic replied:
Would you also remove the same error message from [here](https://github.com/llvm/llvm-project/blob/d648f4ccb8c74060e9e6d75b46f54ac0127e4302/clang/lib/Parse/ParseHLSLRootSignature.cpp#L161)?

They were added just to have _some_ indication that lexing failed.

llvm-beanz replied:
Is it more useful to the user to note a failure in lexing, _or_ is it more useful at the parsing level to be able to recognize that the token was set to invalid when something else was expected?

Clang tends to do the later, which is why you get errors like "expected identifier", because the parser _knows_ the next thing should be an identifier, but the lexer just knows it didn't produce the right result (or any result).

llvm-beanz commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,171 @@
+  return false;
+}
+
+bool RootSignatureLexer::Lex(SmallVector<RootSignatureToken> &Tokens) {
```
I'm not sure this function should exist.

Most modern compilers don't lex entire source files into allocated lists of tokens then parse, they do the two at the same time never allocating more than a couple tokens at a time. You can see this in Clang where the `ConsumeToken` API advances the token to the next token, and is called throughout the parser.

I've seen other compilers where the tokenizer is implemented as effectively an iterator where the iterator state tracks the current token.

For context, the reason compilers tend to do this is because lexing isn't always "fast". In particular translation of numeric literals into actual integer values can be quite slow, so if you lex all the tokens, you do all the transformations even if an error would be detected during parsing. If you lex one token at a time, you stop lexing once the error is encountered providing faster user feedback.

inbelic replied:
Understood, I have updated this to only support a `ConsumeToken`/`PeekNextToken` api.

It could be updated further to define and be used as an iterator. Although from how it is used in the parser (usually through `PeekNextToken`). I think it would essentially be just renaming `ConsumeToken` to the `++` operator. Maybe there are other benefits I don't see?

inbelic committed:
review: update diagnostics message

inbelic committed:
review: update linked todo

inbelic committed:
review: remove api for pre-allocating tokens

- we want to prevent the pre-allocation of all the tokens and rather let
the parser allocate the tokens as it goes along
- this allows for a quicker cycle to the user if there is a parsing
error as we don't need to lex all of the tokens before trying to parse

- as such, this change provides a new api for interacting with the lexer
through ConsumeToken and PeekNextToken as opposed to generated the
entire list of lexed tokens

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,171 @@
+  // Catch this now as the Literal Parser will accept it as valid
+  if (NumSpelling.empty()) {
+    PP.getDiagnostics().Report(Result.TokLoc,
+                               diag::err_hlsl_invalid_number_literal);
```
That is correct. Updated to the concise message

inbelic committed:
clang format

llvm-beanz commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,194 @@
+bool RootSignatureLexer::LexNumber(RootSignatureToken &Result) {
+  // NumericLiteralParser does not handle the sign so we will manually apply it
+  bool Negative = Buffer.front() == '-';
+  bool Signed = Negative || Buffer.front() == '+';
```
I have a grammar question here. Is the `-` or `+` part of the numeric literal token, or is it better (as in C++), that the unary operator be its own token? The reason I ask, is that if you look at the C++ (& HLSL) grammar leading `-` is a unary operator, _not_ part of the numeric literal token string. This is why the Clang parser really only supports parsing positive numbers, the negation is a modifier.

Where this becomes relevant is a few lines below where you emit the `err_hlsl_expected_number_literal`, that's really an error that should be coming from _parsing_ not lexing because you're combining adjacent tokens and recognizing that the grammar requires a number follow the sign token.

inbelic replied:
Updated to have seperate tokens for the `+` and `-` and to defer this parsing for the parser.

llvm-beanz commented on clang/include/clang/Basic/DiagnosticLexKinds.td:
```diff
@@ -1017,4 +1017,15 @@ Error<"'#pragma unsafe_buffer_usage' was not ended">;
+let CategoryName = "Root Signature Lexical Issue" in {
+  def err_hlsl_expected_number_literal: Error<"expected numberic literal">;
+  def err_hlsl_number_literal_overflow :
+    Error<"integer literal is too large to be represented in any %select{signed |}0 integer type">;
```
I know I suggested matching the language in the other clang diagnostics, but for root signatures these are required to be 32-bit types, not _any_ integer type.
```suggestion
    Error<"integer literal is too large to be represented in 32-bit %select{signed |}0 integer type">;
```

llvm-beanz commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,194 @@
+  return isdigit(C); // integer support
+}
+
+bool RootSignatureLexer::LexNumber(RootSignatureToken &Result) {
```
As another note, lexing would usually identify the token, not try to interpret it, meaning any string of characters that meet the definition of a numeric constant. For example in the following code:

```
int8_t C = 2048;
float F = -2.8;
```

The token sequence generated by the lexer is [ `int8_t`, `C`, `=`, `2048`, `;`, `float`, `F`, `=` ,`-`, `2.8`, `;`].
Lexing does not validate that 2048 cannot fit into C.

I think most of what this function does is more accurately "parsing" a number rather than lexing.

llvm-beanz replied:
What did you do to call this resolved? It looks like you're still parsing the number?

llvm-beanz commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,194 @@
+  if (EndOfBuffer()) {
+    // Report unexpected end of tokens error
+    PP.getDiagnostics().Report(SourceLoc,
+                               diag::err_hlsl_rootsig_unexpected_eos);
```
This is similar to the case above. Is it better for the lexer to report this, or to surface an end_of_stream token and have the parser report a more context-aware error? I tend to think the later.

llvm-beanz reviewed (commented):
In general I think this is very much moving in the right direction. I have two high level areas of feedback (see my more specific comments inline).

In general, Lexer's don't often have the context to provide actionable error messages. As a result it is _usually_ better for the Lexer to push up an invalid token to the parser and allow the parser to provide a more contextually relevant diagnostic.

The other area is around what is the responsibility of the parser vs the lexer. In general lexers just break the input into tokens, the parser then interprets and validates the tokens. In particular around numbers this is doing a lot of interpretation of the token values. I think it is probably better to have a clearer separation.

inbelic committed:
review: fix some typos

inbelic committed:
review: remove unique end_of_stream lexing error

- instead of reporting the unique end_of_stream error, we can instead
pass up an end_of_stream token and let the unexpected diag reporting
handles this in the parser with more context

inbelic committed:
review: remove unique invalid_token lexing error

- similarily, we can let the unexpected_token diagnostics reporting
handle this error by passing up an invalid token and provide more
context

inbelic committed:
review: punt parsing of a signed integer to the parser

inbelic committed:
self-review: define an error token to denote lex errors during peek

- the current use of invalid is not able to distinguish between a lex
error that has been reported to diagnostics and an invalid token that
should be reported during parsing
- an error token is defined as the default token to denote a lex error
that has been reported and shouldn't be handled by the parsing

llvm-beanz commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -0,0 +1,166 @@
+
+bool RootSignatureLexer::LexNumber(RootSignatureToken &Result) {
+  // Retrieve the possible number
+  StringRef NumSpelling = Buffer.take_while(IsNumberChar);
```
If I'm reading this correctly right now if someone writes `1.3` the parser will break the token at the `.`, call it an integer literal and return `1` for the value. I assume that's not the expected behavior?

inbelic replied:
Correct. It is the expected behaviour for this pr, as we are not expecting to handle floats until https://github.com/llvm/llvm-project/issues/126565. This was split out because I want to have a pr that will have dedicated documentation and testing to show that the float parsing is equivalent to DXC. Which I think would get bogged down and increase the scope when including it in this pr.

inbelic committed:
review: defer handling of numeric constants to the parser -> remove error states

- as commented, using the `NumericLiteralParser` during lexing of
numeric constants falls more under parsing rather than lexing. this
commit removes the use of this completely from the lexer

- this implies that (after moving the register error to parsing) there
are no longer any errors to report during lexing. instead we will just
create an `invalid` or `end_of_stream` token and let the parser provide
an error with more context

- thus we change the api of `ConsumeToken`/`PeekNextToken` so that it no
longer needs to denote an error and will just return a token. this
greatly simplifies the lexer implementation as we no longer need to pass
down the pre-processor

llvm-beanz commented on clang/include/clang/Parse/ParseHLSLRootSignature.h:
```diff
@@ -0,0 +1,89 @@
+#ifndef LLVM_CLANG_PARSE_PARSEHLSLROOTSIGNATURE_H
+#define LLVM_CLANG_PARSE_PARSEHLSLROOTSIGNATURE_H
+
+#include "clang/AST/APValue.h"
```
Is this needed?

In general it is also best to prefer forward declarations in headers rather than including headers where possible to reduce compile time overhead.

inbelic replied:
Removed this and the other no longer needed headers

llvm-beanz reviewed (commented):
My last real question here is whether this belongs in clang/Parse or if it should be in clang/Lex. Since it is now strictly Lexing it shouldn't have any dependencies on the parse or AST libraries right?

inbelic committed:
review: move to lex dir

inbelic committed:
review: clean up after moving

- remove uneeded headers
- rename to new directory

inbelic committed:
self-review: whoops missed renaming header guards

llvm-beanz commented on clang/include/clang/Lex/HLSLRootSignatureTokenKinds.def:
```diff
@@ -0,0 +1,124 @@
+//===--- HLSLRootSignature.def - Tokens and Enum Database -------*- C++ -*-===//
+
+//
```
nit:
```suggestion
//
```

llvm-beanz reviewed (approved):
One small formatting cleanup in a suggestion. LGTM.

inbelic committed:
Update clang/include/clang/Lex/HLSLRootSignatureTokenKinds.def

Co-authored-by: Chris B <beanz@abolishcrlf.org>
